apps/src/lab2/views/ExtraLinks.tsx

```diff
@@ -14,61 +15,91 @@ interface PermissionResponse {
+  permissions: string[];
+}
+
+async function fetchPermissionsAndLinkData(
+  levelId: number,
+  channelId?: string
+): Promise<PermissionsAndLinkData> {
+  // Fetch permissions.
+  const permissionsResponse = await HttpClient.fetchJson<PermissionResponse>(
+    '/api/v1/users/current/permissions'
+  );
+  const {permissions} = permissionsResponse.value;
+
+  // Fetch level link data.
+  let levelLinkData: ExtraLinksLevelData | undefined;
+  if (permissions.includes(PERMISSIONS.LEVELBUILDER)) {
+    const levelLinkDataResponse =
+      await HttpClient.fetchJson<ExtraLinksLevelData>(
+        `/levels/${levelId}/extra_links`
+      );
+    levelLinkData = levelLinkDataResponse.value;
+  }
+
+  // Fetch project link data.
+  let projectLinkData: ExtraLinksProjectData | undefined;
+  if (permissions.includes(PERMISSIONS.PROJECT_VALIDATOR)) {
+    const levelProjectDataResponse =
+      await HttpClient.fetchJson<ExtraLinksProjectData>(
+        `/projects/${channelId}/extra_links`
+      );
+    projectLinkData = levelProjectDataResponse.value;
+  }
+
+  // Return fetched permissions and link data.
+  return {
+    levelLinkData,
+    projectLinkData,
+    permissions,
+  };
 }
 
 // If the user has permission to see extra links, fetch extra links for the level,
 // then display a modal with the link data.
 const ExtraLinks: React.FunctionComponent<ExtraLinksProps> = ({
   levelId,
 }: ExtraLinksProps) => {
-  const {loading, data} = useFetch('/api/v1/users/current/permissions');
+  const [isLoading, setIsLoading] = useState(false);
+  const [PermissionsAndLinkData, setPermissionsAndLinkData] =
+    useState<PermissionsAndLinkData | null>(null);
   const [isModalOpen, setIsModalOpen] = useState(false);
-  const [linkData, setLinkData] = useState<ExtraLinksResponse | null>(null);
 
-  useEffect(() => {
-    const permissionData = data
-      ? (data as PermissionResponse)
-      : {permissions: []};
-    if (permissionData.permissions.includes(PERMISSIONS.LEVELBUILDER)) {
-      try {
-        HttpClient.fetchJson<ExtraLinksResponse>(
-          `/levels/${levelId}/extra_links`
-        ).then(response => {
-          setLinkData(response.value);
-        });
-      } catch (e) {
-        console.error('Error fetching extra links', e);
-      }
-    }
-  }, [data, levelId]);
-
-  const permissionData = data
-    ? (data as PermissionResponse)
-    : {permissions: []};
+  const channelId = useAppSelector(
+    state => state.lab.channel && state.lab.channel.id
+  );
 
-  if (!permissionData.permissions.includes(PERMISSIONS.LEVELBUILDER)) {
+  useEffect(() => {
+    setIsLoading(true);
+    fetchPermissionsAndLinkData(levelId, channelId).then(data => {
+      setPermissionsAndLinkData(data);
+      setIsLoading(false);
+    });
+  }, [levelId, channelId]);
+  const {levelLinkData, projectLinkData, permissions} =
+    PermissionsAndLinkData || {};
+  if (
+    !permissions?.includes(PERMISSIONS.LEVELBUILDER) &&
```

Comment:
Actually, do we need to check the permissions at all? Could this just be
```
if (loading || permissionsAndLinkData === undefined) {
  return null;
}

const {levelLinkData, projectLinkData} = permissionsAndLinkData;
if (levelLinkData === undefined && projectLinkData === undefined) {
  return null;
}

// if we've gotten here, we know we have at least one type of data and some permissions to view it
```
In which case we don't even need to return the permissions and could just call it `extraLinksData` or something

Reply:
Oh nice! And can simplify even more with one condition:
```
return isLoading || (!levelLinkData && !projectLinkData) ? null : (
  <>
...
```

Thanks for all the refactor suggestions!